### PR TITLE
[DC-861] Add support for empty string to SearchConcepts

### PR DESCRIPTION
**Pull request by rjohanek**
Background:
In a recent PR Evan discovered searchConcepts does not support the empty string being passed in.  https://github.com/DataBiosphere/terra-ui/pull/4660 
This functionality is needed for the loading of the initial search page.

Changes:
This PR adds support for empty string in searchConcepts endpoint by only rendering the search clauses if the string is not empty. 

Testing:
I added a test of the searchConceptsQueryBuilder (where the change is implemented).
I also tested the newly generated query in BigQuery.

Question:
Do I need to make the searchText field not required for the API in order to support the empty string?

### CODE REVIEW

File: src/main/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilder.java [hunk 1]
```diff
@@ -31,33 +31,43 @@ public static String buildSearchConceptsQuery(
...
+      Query query =
+          new Query(List.of(nameField, idField), List.of(conceptTableVariable), domainClause, 100);
+      return query.renderSQL(platform);
+    } else {
 
-    // search concept name clause filters for the search text based on field concept_code
-    var searchCodeClause =
-        createSearchConceptClause(
-            conceptTablePointer, conceptTableVariable, searchText, "concept_code");
+      // search concept name clause filters for the search text based on field concept_name
```
Comment (evansuslovich) [inline src/main/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilder.java]: what's the difference between `searchCodeClause` and `searchNameClause`?

File: src/test/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilderTest.java [hunk 2]
```diff
@@ -27,15 +27,28 @@ void buildSearchConceptsQuery(CloudPlatform platform) {
...
                 + "AND (CONTAINS_SUBSTR(c.concept_name, 'cancer') "
                 + "OR CONTAINS_SUBSTR(c.concept_code, 'cancer'))) "
                 + "LIMIT 100"));
   }
 
+  @ParameterizedTest
+  @EnumSource(CloudPlatform.class)
+  void buildSearchConceptsQueryEmpty(CloudPlatform platform) {
+    assertThat(
+        "generated SQL for empty search string is correct",
```
Comment (evansuslovich) [inline src/test/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilderTest.java:43]: great assert statement!

File: src/main/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilder.java (hunk 1 above)
Comment (rjohanek) [reply src/main/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilder.java]: searchName clause adds a string of sequel that searches for matches in the concept_name field while searchCode clause does the same for searches against the concept_code field. Together his allows users to search by either concept name or concept code!

File: src/main/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilder.java [hunk 3]
```diff
@@ -31,33 +31,43 @@ public static String buildSearchConceptsQuery(
...
-    var searchNameClause =
-        createSearchConceptClause(
-            conceptTablePointer, conceptTableVariable, searchText, "concept_name");
+    // if the search test is empty do not include the search clauses
+    // return all concepts in the specified domain
+    if (searchText.isEmpty()) {
+      Query query =
+          new Query(List.of(nameField, idField), List.of(conceptTableVariable), domainClause, 100);
+      return query.renderSQL(platform);
+    } else {
```
Comment (rjohanek) [inline src/main/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilder.java]: this is the only functional change the rest is formatting

**Review by evansuslovich — APPROVED** [review]
Awesome work!

File: src/test/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilderTest.java [hunk 4]
```diff
@@ -27,15 +27,28 @@ void buildSearchConceptsQuery(CloudPlatform platform) {
     assertThat(
         "generated SQL is correct",
         SearchConceptsQueryBuilder.buildSearchConceptsQuery(
-            "condition", "cancer", s -> s, CloudPlatformWrapper.of(platform)),
+            "Condition", "cancer", s -> s, CloudPlatformWrapper.of(platform)),
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilderTest.java]: Is this change necessary? The test data doesn't need to be the same as the real data.

File: src/main/resources/api/data-repository-openapi.yaml [hunk 5]
```diff
@@ -4601,10 +4601,10 @@ paths:
           required: true
           schema:
             type: string
-        - in: path
+        - in: query
           description: User specified text to search concepts for.
           name: searchText
-          required: true
+          required: false
```
Comment (rjohanek) [inline src/main/resources/api/data-repository-openapi.yaml]: instead of having to pass in the empty string for the initial render on the UI, nothing needs to be passed in

**Review by snf2ye — APPROVED** [review]
LGTM - thanks for walking through the changes with me.

File: src/test/java/bio/terra/app/controller/DatasetsApiControllerTest.java [hunk 6]
```diff
@@ -441,14 +441,21 @@ void getSnapshotBuilderCount() throws Exception {
     verifyAuthorizationCall(IamAction.VIEW_SNAPSHOT_BUILDER_SETTINGS);
   }
 
-  @Test
-  void testSearchConcepts() throws Exception {
+  private static Stream<String> searchTextArguments() {
+    return Stream.of("cancer", "", null);
+  }
```
Comment (snf2ye) [inline src/test/java/bio/terra/app/controller/DatasetsApiControllerTest.java:446]: Look at the parameterized test! 🥳

File: src/test/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilderTest.java [hunk 7]
```diff
@@ -27,15 +27,28 @@ void buildSearchConceptsQuery(CloudPlatform platform) {
...
   }
 
+  @ParameterizedTest
+  @EnumSource(CloudPlatform.class)
+  void buildSearchConceptsQueryEmpty(CloudPlatform platform) {
+    assertThat(
+        "generated SQL for empty search string is correct",
+        SearchConceptsQueryBuilder.buildSearchConceptsQuery(
+            "Condition", "", s -> s, CloudPlatformWrapper.of(platform)),
+        equalToCompressingWhiteSpace(
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilderTest.java:46]: I wonder if this should really be `equalToIgnoringWhiteSpace`? It works by compressing the whitespace but that detail isn't important to the caller

File: src/main/resources/api/data-repository-openapi.yaml (hunk 5 above)
Comment (pshapiro4broad) [reply src/main/resources/api/data-repository-openapi.yaml]: I think `required: false` is the default so it can now be omitted.

File: src/main/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilder.java [hunk 8]
```diff
@@ -31,33 +31,44 @@ public static String buildSearchConceptsQuery(
...
-        createSearchConceptClause(
-            conceptTablePointer, conceptTableVariable, searchText, "concept_name");
+    // if the search test is empty do not include the search clauses
+    // return all concepts in the specified domain
+    if (searchText == null || searchText.isEmpty()) {
+      // TODO: DC-845 Implement pagination, remove hardcoded limit
+      Query query =
+          new Query(List.of(nameField, idField), List.of(conceptTableVariable), domainClause, 100);
+      return query.renderSQL(platform);
+    } else {
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilder.java]: Using `else` after a `return` is unnecessary since the `return` has already interrupted the normal flow of execution.
```suggestion
      return query.renderSQL(platform);
    }
```
Either way works but, for me, omitting `else` makes it clearer.

**Review by pshapiro4broad — APPROVED** [review]

File: src/test/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilderTest.java (hunk 4 above)
Comment (rjohanek) [reply src/test/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilderTest.java]: No it's not necessary

File: src/test/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilderTest.java (hunk 7 above)
Comment (rjohanek) [reply src/test/java/bio/terra/service/snapshotbuilder/utils/SearchConceptsQueryBuilderTest.java:46]: The IDE tells me equalToIgnoringWhiteSpace is deprecated and to use equalToCompressingWhiteSpace in its place
